pkg/apiserver/storage/ram/store.go

```diff
@@ -266,10 +275,46 @@ func (s *store) dispatchEvents() {
 }
 
 func (s *store) dispatchEvent(event antreastorage.InternalEvent) {
-	s.watcherMutex.Lock()
-	defer s.watcherMutex.Unlock()
-	// TODO: Optimize this to dispatch the event based on watchers' selector.
-	for _, watcher := range s.watchers {
-		watcher.add(event)
+	var failedWatchers []*storeWatcher
+
+	func() {
+		s.watcherMutex.RLock()
+		defer s.watcherMutex.RUnlock()
+
+		// First try to send events without blocking, to avoid setting up a timer
+		// for every event.
+		// blockedWatchers keeps watchers whose buffer are full.
+		var blockedWatchers []*storeWatcher
+		// TODO: Optimize this to dispatch the event based on watchers' selector.
+		for _, watcher := range s.watchers {
+			if !watcher.nonBlockingAdd(event) {
+				blockedWatchers = append(blockedWatchers, watcher)
+			}
+		}
+		if len(blockedWatchers) == 0 {
+			return
+		}
+		klog.V(2).Infof("%d watchers were not available to receive event %+v immediately", len(blockedWatchers), event)
+
+		// Then try to send events to blocked watchers with a timeout. If it
+		// timeouts, it means the watcher is too slow to consume the events or the
+		// underlying connection is already dead, terminate the watcher in this case.
+		timer := time.NewTimer(watcherAddTimeout)
+		defer timer.Stop()
+
+		for _, watcher := range blockedWatchers {
+			if !watcher.add(event, timer) {
+				failedWatchers = append(failedWatchers, watcher)
+				// setting timer to nil to let watcher know know the timer has fired.
+				timer = nil
+			}
+		}
+	}()
+
+	// Terminate unresponsive watchers, this must be executed without watcherMutex as
+	// watcher.Stop will require the lock itself.
+	for _, watcher := range failedWatchers {
+		klog.Warningf("Forcing stopping watcher (selectors: %v) due to unresponsiveness", watcher.selectors)
+		watcher.Stop()
```

Comment:
out of curiosity: if the antrea-agent is unresponsive temporarily and the corresponding watcher is stopped, will there be a new call to `Watch` eventually?

Reply:
Yes, clients will reconnect the server on disconnection: https://github.com/vmware-tanzu/antrea/blob/60523f5c24c2b833a2ceac45dc6668a122a6f681/pkg/agent/controller/networkpolicy/networkpolicy_controller.go#L115-L117

Reply:
Thanks for the pointer!

Reply:
Could we add this to comments?

Reply:
Done